File: py/multitenant/subdomain.py
"""Subdomain-based tenant resolution utilities.

Handles parsing and validating subdomain.domain.app format.
"""

import re
from dataclasses import dataclass
from typing import Optional


@dataclass
class SubdomainConfig:
    """Configuration for subdomain-based multitenancy."""
    
    # Base domain (e.g., "domain.app")
    base_domain: str
    # Whether to allow www as a valid subdomain
    allow_www: bool = False
    # Reserved subdomains that cannot be used by tenants
    reserved_subdomains: tuple[str, ...] = (
        "www",
        "api",
        "admin",
        "app",
        "auth",
        "login",
        "signup",
        "static",
        "assets",
        "cdn",
        "mail",
        "smtp",
        "ftp",
        "help",
        "support",
        "docs",
        "status",
    )
    # Minimum subdomain length
    min_length: int = 2
    # Maximum subdomain length
    max_length: int = 63
    # Pattern for valid subdomains
    pattern: str = r"^[a-z0-9]([a-z0-9-]*[a-z0-9])?$"


@dataclass
class ParsedHost:
    """Result of parsing a host into its components."""
    
    subdomain: Optional[str]
    base_domain: str
    is_valid: bool
    is_platform: bool  # True if this is the main platform (no subdomain)
    error: Optional[str] = None


def parse_host(host: str, config: SubdomainConfig) -> ParsedHost:
    """Parse a host string into subdomain and base domain.
    
    Examples:
        - "tenant1.domain.app" -> subdomain="tenant1", base_domain="domain.app"
        - "domain.app" -> subdomain=None, is_platform=True
        - "www.domain.app" -> subdomain=None (if allow_www) or error
    """
    # Normalize host
    host = host.lower().strip()
    
    # Remove port if present
    if ":" in host:
        host = host.split(":")[0]
    
    base_domain = config.base_domain.lower()
    
    # Check if it's the base domain (platform level)
    if host == base_domain:
        return ParsedHost(
            subdomain=None,
            base_domain=base_domain,
            is_valid=True,
            is_platform=True,
        )
    
    # Check if host ends with base domain
    if not host.endswith(f".{base_domain}"):
        return ParsedHost(
            subdomain=None,
            base_domain=base_domain,
            is_valid=False,
            is_platform=False,
            error=f"Host '{host}' does not match base domain '{base_domain}'",
        )
    
    # Extract subdomain
    subdomain = host[: -(len(base_domain) + 1)]
    
    # Handle www
    if subdomain == "www":
        if config.allow_www:
            return ParsedHost(
                subdomain=None,
                base_domain=base_domain,
                is_valid=True,
                is_platform=True,
            )
        else:
            return ParsedHost(
                subdomain=None,
                base_domain=base_domain,
                is_valid=False,
                is_platform=False,
                error="www subdomain is not allowed",
            )
    
    # Validate subdomain
    validation_error = validate_subdomain(subdomain, config)
    if validation_error:
        return ParsedHost(
            subdomain=subdomain,
            base_domain=base_domain,
            is_valid=False,
            is_platform=False,
            error=validation_error,
        )
    
    return ParsedHost(
        subdomain=subdomain,
        base_domain=base_domain,
        is_valid=True,
        is_platform=False,
    )


def validate_subdomain(subdomain: str, config: SubdomainConfig) -> Optional[str]:
    """Validate a subdomain string. Returns error message or None if valid."""
    
    if not subdomain:
        return "Subdomain cannot be empty"
    
    if len(subdomain) < config.min_length:
        return f"Subdomain must be at least {config.min_length} characters"
    
    if len(subdomain) > config.max_length:
        return f"Subdomain cannot exceed {config.max_length} characters"
    
    if subdomain in config.reserved_subdomains:
        return f"Subdomain '{subdomain}' is reserved"
    
    if not re.match(config.pattern, subdomain):
        return "Subdomain must contain only lowercase letters, numbers, and hyphens"
    
    if subdomain.startswith("-") or subdomain.endswith("-"):
        return "Subdomain cannot start or end with a hyphen"
    
    return None


def build_tenant_url(subdomain: str, config: SubdomainConfig, scheme: str = "https") -> str:
    """Build a full URL for a tenant subdomain."""
    return f"{scheme}://{subdomain}.{config.base_domain}"


def get_available_subdomains(
    requested: str,
    existing: set[str],
    config: SubdomainConfig,
    count: int = 5,
) -> list[str]:
    """Suggest available subdomains based on a requested name.
    
    Useful for signup flows when the requested subdomain is taken.
    """
    suggestions = []
    base = re.sub(r"[^a-z0-9]", "", requested.lower())
    
    if not base:
        base = "team"
    
    # Try the base name first
    if base not in existing and validate_subdomain(base, config) is None:
        suggestions.append(base)
    
    # Try with numbers
    for i in range(1, 100):
        candidate = f"{base}{i}"
        if candidate not in existing and validate_subdomain(candidate, config) is None:
            suggestions.append(candidate)
            if len(suggestions) >= count:
                break
    
    return suggestions

File: js/src/subdomain.ts
/**
 * Subdomain utilities for multitenant applications.
 * Handles parsing and validating subdomain.domain.app format.
 */

export interface SubdomainConfig {
  /** Base domain (e.g., "domain.app") */
  baseDomain: string;
  /** Whether to allow www as a valid subdomain */
  allowWww?: boolean;
  /** Reserved subdomains that cannot be used by tenants */
  reservedSubdomains?: string[];
  /** Minimum subdomain length */
  minLength?: number;
  /** Maximum subdomain length */
  maxLength?: number;
}

export interface ParsedHost {
  subdomain: string | null;
  baseDomain: string;
  isValid: boolean;
  isPlatform: boolean;
  error?: string;
}

const DEFAULT_RESERVED_SUBDOMAINS = [
  'www',
  'api',
  'admin',
  'app',
  'auth',
  'login',
  'signup',
  'static',
  'assets',
  'cdn',
  'mail',
  'smtp',
  'ftp',
  'help',
  'support',
  'docs',
  'status',
];

const SUBDOMAIN_PATTERN = /^[a-z0-9]([a-z0-9-]*[a-z0-9])?$/;

/**
 * Parse a host string into subdomain and base domain.
 */
export function parseHost(host: string, config: SubdomainConfig): ParsedHost {
  const {
    baseDomain,
    allowWww = false,
    reservedSubdomains = DEFAULT_RESERVED_SUBDOMAINS,
    minLength = 2,
    maxLength = 63,
  } = config;

  // Normalize host
  let normalizedHost = host.toLowerCase().trim();

  // Remove port if present
  if (normalizedHost.includes(':')) {
    normalizedHost = normalizedHost.split(':')[0];
  }

  const normalizedBase = baseDomain.toLowerCase();

  // Check if it's the base domain (platform level)
  if (normalizedHost === normalizedBase) {
    return {
      subdomain: null,
      baseDomain: normalizedBase,
      isValid: true,
      isPlatform: true,
    };
  }

  // Check if host ends with base domain
  if (!normalizedHost.endsWith(`.${normalizedBase}`)) {
    return {
      subdomain: null,
      baseDomain: normalizedBase,
      isValid: false,
      isPlatform: false,
      error: `Host '${host}' does not match base domain '${baseDomain}'`,
    };
  }

  // Extract subdomain
  const subdomain = normalizedHost.slice(0, -(normalizedBase.length + 1));

  // Handle www
  if (subdomain === 'www') {
    if (allowWww) {
      return {
        subdomain: null,
        baseDomain: normalizedBase,
        isValid: true,
        isPlatform: true,
      };
    } else {
      return {
        subdomain: null,
        baseDomain: normalizedBase,
        isValid: false,
        isPlatform: false,
        error: 'www subdomain is not allowed',
      };
    }
  }

  // Validate subdomain
  const validationError = validateSubdomain(subdomain, {
    reservedSubdomains,
    minLength,
    maxLength,
  });

  if (validationError) {
    return {
      subdomain,
      baseDomain: normalizedBase,
      isValid: false,
      isPlatform: false,
      error: validationError,
    };
  }

  return {
    subdomain,
    baseDomain: normalizedBase,
    isValid: true,
    isPlatform: false,
  };
}

/**
 * Validate a subdomain string.
 */
export function validateSubdomain(
  subdomain: string,
  options: {
    reservedSubdomains?: string[];
    minLength?: number;
    maxLength?: number;
  } = {}
): string | null {
  const {
    reservedSubdomains = DEFAULT_RESERVED_SUBDOMAINS,
    minLength = 2,
    maxLength = 63,
  } = options;

  if (!subdomain) {
    return 'Subdomain cannot be empty';
  }

  if (subdomain.length < minLength) {
    return `Subdomain must be at least ${minLength} characters`;
  }

  if (subdomain.length > maxLength) {
    return `Subdomain cannot exceed ${maxLength} characters`;
  }

  if (reservedSubdomains.includes(subdomain)) {
    return `Subdomain '${subdomain}' is reserved`;
  }

  if (!SUBDOMAIN_PATTERN.test(subdomain)) {
    return 'Subdomain must contain only lowercase letters, numbers, and hyphens';
  }

  if (subdomain.startsWith('-') || subdomain.endsWith('-')) {
    return 'Subdomain cannot start or end with a hyphen';
  }

  return null;
}

/**
 * Build a full URL for a tenant subdomain.
 */
export function buildTenantUrl(
  subdomain: string,
  config: SubdomainConfig,
  scheme: string = 'https'
): string {
  return `${scheme}://${subdomain}.${config.baseDomain}`;
}

/**
 * Get the current subdomain from window.location.
 */
export function getCurrentSubdomain(config: SubdomainConfig): string | null {
  if (typeof window === 'undefined') {
    return null;
  }
  const parsed = parseHost(window.location.host, config);
  return parsed.subdomain;
}

/**
 * Suggest available subdomains based on a requested name.
 */
export function suggestSubdomains(
  requested: string,
  existingSubdomains: Set<string>,
  config: SubdomainConfig,
  count: number = 5
): string[] {
  const suggestions: string[] = [];
  const base = requested.toLowerCase().replace(/[^a-z0-9]/g, '') || 'team';

  // Try the base name first
  if (!existingSubdomains.has(base) && !validateSubdomain(base, config)) {
    suggestions.push(base);
  }

  // Try with numbers
  for (let i = 1; i < 100 && suggestions.length < count; i++) {
    const candidate = `${base}${i}`;
    if (!existingSubdomains.has(candidate) && !validateSubdomain(candidate, config)) {
      suggestions.push(candidate);
    }
  }

  return suggestions;
}
